File: src/main/java/com/back/mozu/domain/reservation/repository/ReservationRepository.java
package com.back.mozu.domain.reservation.repository;

import com.back.mozu.domain.reservation.entity.Reservation;
import com.back.mozu.domain.reservation.entity.ReservationStatus;
import com.back.mozu.domain.reservation.entity.TimeSlot;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface ReservationRepository extends JpaRepository<Reservation, UUID> {
    Optional<Reservation> findByUserIdAndStatus(UUID customerId, ReservationStatus status);
    boolean existsByUserIdAndTimeSlotAndStatusNot(UUID customerId, TimeSlot timeSlot, ReservationStatus status);
    List<Reservation> findAllByUserId(UUID customerId);
    int countByTimeSlot_Date(LocalDate date);
    Page<Reservation> findAllWithFilters(LocalDate date, LocalTime time, String status, Pageable pageable);

    // 추가
    int countByUserIdAndStatusAndCancelledAtAfter(UUID userId, ReservationStatus status, LocalDateTime dateTime);
    List<Reservation> findAllByStatusAndReleaseAtBefore(ReservationStatus status, LocalDateTime dateTime);
    List<Reservation> findAllByStatusAndCreatedAtBefore(ReservationStatus status, LocalDateTime createdAt);
    List<Reservation> findAllByStatus(ReservationStatus status);
    List<Reservation> findByTimeSlotIdAndStatusOrderByCreatedAt(UUID timeSlotId, ReservationStatus status);
}

File: src/main/java/com/back/mozu/domain/reservation/repository/ReservationRepository.kt
package com.back.mozu.domain.reservation.repository

import com.back.mozu.domain.reservation.entity.Reservation
import com.back.mozu.domain.reservation.entity.ReservationStatus
import com.back.mozu.domain.reservation.entity.TimeSlot
import jakarta.persistence.LockModeType
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Lock
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.util.UUID

interface ReservationRepository : JpaRepository<Reservation, UUID> {
    fun findByUserIdAndStatus(userId: UUID, status: ReservationStatus): Reservation?
    fun existsByUserIdAndTimeSlotAndStatusNot(userId: UUID, timeSlot: TimeSlot, status: ReservationStatus): Boolean
    fun findAllByUserId(userId: UUID): List<Reservation>
    fun countByTimeSlot_Date(date: LocalDate): Int
    fun findAllWithFilters(date: LocalDate?, time: LocalTime?, status: String?, pageable: Pageable): Page<Reservation>
    fun countByUserIdAndStatusAndCancelledAtAfter(userId: UUID, status: ReservationStatus, dateTime: LocalDateTime): Int
    fun findAllByStatusAndReleaseAtBefore(status: ReservationStatus, dateTime: LocalDateTime): List<Reservation>
    fun findAllByStatusAndCreatedAtBefore(status: ReservationStatus, createdAt: LocalDateTime): List<Reservation>
    fun findAllByStatus(status: ReservationStatus): List<Reservation>
    fun findByTimeSlotIdAndStatusOrderByCreatedAt(timeSlotId: UUID, status: ReservationStatus): List<Reservation>

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("SELECT r FROM Reservation r WHERE r.id = :id")
    fun findByIdWithLock(@Param("id") id: UUID): Reservation?
}